### ISSUE TITLE
Description in perlport of the meaning of \n

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9670]
Migrated from [rt.perl.org#63620](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=63620) (status was 'resolved')

Searchable as RT63620$

**Comment by p5pRT** [comment]
### From mail@gunnar.cc


This is a bug report for perl from mail@&#8203;gunnar\.cc\,
generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.




---
The second para in the Newlines section of perlport\.pod describes
the meaning of \\n\. As regards "DOSish perls"\, the description is
not clear IMO\. In a discussion at comp\.lang\.perl\.misc\, Ben Morrow
suggested a changed wording&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- /usr/lib/perl5/5.10.0/pod/perlport.pod      2008-04-18 22:48:31.000000000 -0400
+++ perlport.pod        2009-03-03 13:21:21.000000000 -0500
@@ -88,11 +88,11 @@

 Perl uses C<\n> to represent the "logical" newline, where what is
 logical may depend on the platform in use.  In MacPerl, C<\n> always
-means C<\015>.  In DOSish perls, C<\n> usually means C<\012>, but
-when accessing a file in "text" mode, STDIO translates it to (or
-from) C<\015\012>, depending on whether you're reading or writing.
-Unix does the same thing on ttys in canonical mode.  C<\015\012>
-is commonly referred to as CRLF.
+means C<\015>.  In DOSish perls, C<\n> means C<\012>, but when
+accessing a file, by default a :crlf layer is pushed which will
+result in C<\015\012> appearing in the file.  Unix does the same
+thing on ttys in canonical mode.  C<\015\012> is commonly referred to
+as CRLF.

 To trim trailing newlines from text lines use chomp().  With default
 settings that function looks for a trailing C<\n> character and thus

-----------------------------------------------------------------
---

```
</details>
Flags:
&nbsp;    category=docs
&nbsp;    severity=medium


---
Site configuration information for perl 5\.10\.0&#8203;:

Configured by root at Sat Apr 19 00&#8203;:09&#8203;:44 EDT 2008\.

Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.12\.4\-vs2\.0\, archname=i686\-linux
&nbsp;    uname='linux gunnar\.cc 2\.6\.12\.4\-vs2\.0 \#9 smp fri nov 11 23&#8203;:37&#8203;:46 est 2005 i686 i686 i386 gnulinux '
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-I/usr/include/gdbm'
&nbsp;    ccversion=''\, gccversion='3\.3\.2 20031022 \(Red Hat Linux 3\.3\.2\-1\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =''
&nbsp;    libpth=/usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=/lib/libc\-2\.3\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.2'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.10\.0&#8203;:
&nbsp;    /usr/lib/perl5/5\.10\.0/i686\-linux
&nbsp;    /usr/lib/perl5/5\.10\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.10\.0/i686\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.10\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.1
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.10\.0&#8203;:
&nbsp;    HOME=/home/gunnarh
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/kerberos/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/home/gunnarh/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Mar 03\, 2009 at 10&#8203;:56&#8203;:55AM \-0800\, Gunnar Hjalmarsson wrote&#8203;:

> The second para in the Newlines section of perlport\.pod describes
> the meaning of \\n\. As regards "DOSish perls"\, the description is
> not clear IMO\. In a discussion at comp\.lang\.perl\.misc\, Ben Morrow
> suggested a changed wording&#8203;:
> 
> \-\-\- /usr/lib/perl5/5\.10\.0/pod/perlport\.pod      2008\-04\-18 22&#8203;:48&#8203;:31\.000000000 \-0400
> \+\+\+ perlport\.pod        2009\-03\-03 13&#8203;:21&#8203;:21\.000000000 \-0500
> @&#8203;@&#8203; \-88\,11 \+88\,11 @&#8203;@&#8203;
> 
>  Perl uses C\<\\n> to represent the "logical" newline\, where what is
>  logical may depend on the platform in use\.  In MacPerl\, C\<\\n> always
> \-means C\<\\015>\.  In DOSish perls\, C\<\\n> usually means C\<\\012>\, but
> \-when accessing a file in "text" mode\, STDIO translates it to \(or
> \-from\) C\<\\015\\012>\, depending on whether you're reading or writing\.
> \-Unix does the same thing on ttys in canonical mode\.  C\<\\015\\012>
> \-is commonly referred to as CRLF\.
> \+means C\<\\015>\.  In DOSish perls\, C\<\\n> means C\<\\012>\, but when
> \+accessing a file\, by default a :crlf layer is pushed which will
> \+result in C\<\\015\\012> appearing in the file\.  Unix does the same
> \+thing on ttys in canonical mode\.  C\<\\015\\012> is commonly referred to
> \+as CRLF\.
> 
>  To trim trailing newlines from text lines use chomp\(\)\.  With default
>  settings that function looks for a trailing C\<\\n> character and thus

I find that this new text is less clear than the original\.  In particular\,
what does it mean for C\<\\015\\012> to "appear in a file" that is being read\,
not written?

Ronald

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From mail@gunnar.cc

Ronald J Kimball via RT wrote&#8203;:
> On Tue\, Mar 03\, 2009 at 10&#8203;:56&#8203;:55AM \-0800\, Gunnar Hjalmarsson wrote&#8203;:
> 
>> The second para in the Newlines section of perlport\.pod describes
>> the meaning of \\n\. As regards "DOSish perls"\, the description is
>> not clear IMO\. In a discussion at comp\.lang\.perl\.misc\, Ben Morrow
>> suggested a changed wording&#8203;:
>>
>> \-\-\- /usr/lib/perl5/5\.10\.0/pod/perlport\.pod      2008\-04\-18 22&#8203;:48&#8203;:31\.000000000 \-0400
>> \+\+\+ perlport\.pod        2009\-03\-03 13&#8203;:21&#8203;:21\.000000000 \-0500
>> @&#8203;@&#8203; \-88\,11 \+88\,11 @&#8203;@&#8203;
>>
>>  Perl uses C\<\\n> to represent the "logical" newline\, where what is
>>  logical may depend on the platform in use\.  In MacPerl\, C\<\\n> always
>> \-means C\<\\015>\.  In DOSish perls\, C\<\\n> usually means C\<\\012>\, but
>> \-when accessing a file in "text" mode\, STDIO translates it to \(or
>> \-from\) C\<\\015\\012>\, depending on whether you're reading or writing\.
>> \-Unix does the same thing on ttys in canonical mode\.  C\<\\015\\012>
>> \-is commonly referred to as CRLF\.
>> \+means C\<\\015>\.  In DOSish perls\, C\<\\n> means C\<\\012>\, but when
>> \+accessing a file\, by default a :crlf layer is pushed which will
>> \+result in C\<\\015\\012> appearing in the file\.  Unix does the same
>> \+thing on ttys in canonical mode\.  C\<\\015\\012> is commonly referred to
>> \+as CRLF\.
>>
>>  To trim trailing newlines from text lines use chomp\(\)\.  With default
>>  settings that function looks for a trailing C\<\\n> character and thus
> 
> I find that this new text is less clear than the original\.  In particular\,
> what does it mean for C\<\\015\\012> to "appear in a file" that is being read\,
> not written?

Ok\, I see what you mean\. It was taken directly from this message&#8203;:

http&#8203;://groups\.google\.com/group/comp\.lang\.perl\.misc/msg/7d6ca2e5eef111c5

Then how about&#8203;:

In DOSish perls\, C\<\\n> means C\<\\012>\. When reading a file in "text" 
mode\, a :crlf layer is pushed which translates C\<\\015\\012> to C\<\\012>\. 
The reverse translation takes place when saving in "text" mode which 
results in C\<\\015\\012> appearing in the file\.

Best regards\,

Gunnar

**Comment by p5pRT** [comment]
### From @briandfoy

Fixed by commit 51d9476f59194f5f7559ec77351329963b330df8

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
